### PR TITLE
Enable GZip and Deflate compression for binary messages over TCP

### PR DESCRIPTION
**Pull request by jonlouie**
### Description
Closes: #595 

This PR allows server bindings to be configured for GZip and Deflate compression formats for binary messages over TCP.

---

### Testing
* Unit tests show end-to-end communication works as expected
* Manual inspection of network activity showed the messages sent in the unit tests were in fact compressed:
Uncompressed message: `new string('a', 10000)  // string of length 10,000`

Compression | Message size
--- | --- 
Uncompressed | 10198 bytes
GZip |  233 bytes
Deflate | 216 bytes

---

### Concerns/Follow-up questions:
* ~This PR only enables recognition of GZip and Deflate compression formats if the content type is `application/soap+msbinsession1`. It was the only content type I was able to figure out how to test so the content type of `application/soap+msbin1` was left out. Are there any concerns leaving this content type unsupported? If so, what is the best way to test this case?~

### CODE REVIEW

File: src/CoreWCF.NetTcp/src/CoreWCF/Channels/Framing/FramingDecoder.cs [hunk 1]
```diff
@@ -386,6 +386,14 @@ public static string GetString(FramingEncodingType type)
                     return FramingEncodingString.Binary;
                 case FramingEncodingType.BinarySession:
                     return FramingEncodingString.BinarySession;
+                //case FramingEncodingType.ExtendedBinaryGZip:
+                //    return FramingEncodingString.ExtendedBinaryGZip;
```
Comment (mconnew) [inline src/CoreWCF.NetTcp/src/CoreWCF/Channels/Framing/FramingDecoder.cs:390]: I believe these commented out variations are needed to support compression when using TransferMode Streamed as that doesn't maintain a session between requests. So I think you need to include the non-session variants too.

File: src/CoreWCF.NetTcp/src/CoreWCF/Channels/Framing/FramingEncoder.cs [hunk 2]
```diff
@@ -130,7 +130,23 @@ private EncodedContentType(string contentType)
 
         public static EncodedContentType Create(string contentType)
         {
-            if (contentType == FramingEncodingString.BinarySession)
+            if (contentType == FramingEncodingString.ExtendedBinarySessionDeflate)
+            {
+                return new EncodedContentType(FramingEncodingType.ExtendedBinarySessionDeflate);
+            }
+            //else if (contentType == FramingEncodingString.ExtendedBinaryDeflate)
+            //{
```
Comment (mconnew) [inline src/CoreWCF.NetTcp/src/CoreWCF/Channels/Framing/FramingEncoder.cs:138]: Same comment as previous, we need to non-session variations too.

File: src/CoreWCF.NetTcp/src/CoreWCF/Channels/Framing/FramingDecoder.cs (hunk 1 above)
Comment (jonlouie) [reply src/CoreWCF.NetTcp/src/CoreWCF/Channels/Framing/FramingDecoder.cs]: Thanks for pointing out TransferMode.Streamed. I've uncommented those variations and added appropriate test cases.

File: src/CoreWCF.NetTcp/src/CoreWCF/Channels/Framing/FramingEncoder.cs (hunk 2 above)
Comment (jonlouie) [reply src/CoreWCF.NetTcp/src/CoreWCF/Channels/Framing/FramingEncoder.cs]: Uncommented these non-session variants as well

**Comment by jonlouie** [comment]
It looks like this test is failing for `Debug Windows_net5.0` and `Release Linux_net5.0`: CoreWCF.NetTcp.Tests.TransportWithMessageCredentialNetTCPTest.NetTCPRequestReplyWithTransportMessageEchoStringDemuxFailure

@mconnew Any idea if this is a flaky test? I'm not able to reproduce the failure locally.

**Comment by mconnew** [comment]
It's a flakey test.

**Comment by mconnew** [comment]
I'm really starting to get annoyed by that flakey test. I can't reproduce it locally though as my computer is too fast. It's a race condition on a failure case causing the server to return a 500 error instead of the correct fault message. It's a rare edge case for a very narrow scenario and it's causing the wrong failure on the client but isn't causing a passing scenario to fail so I haven't put the time in to try really hard to reproduce locally (need to set up a VM and artificially slow it down). So for now I just rerun the test and it usually passes the next time.

File: src/CoreWCF.NetTcp/tests/CompressionFormatTests.cs [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+                var result = channel.EchoString(testString);
+                Assert.Equal(testString, result);
+                ((System.ServiceModel.IClientChannel)channel).Close();
+            }
+        }
+
+        public static IEnumerable<object[]> GetTestVariations()
+        {
+            yield return new object[]
+            {
```
Comment (mconnew) [inline src/CoreWCF.NetTcp/tests/CompressionFormatTests.cs:62]: You can simplify this by using loops. You can also cast an enum between types as they are backed by an integer. You could do something like this:
```c#
foreach (CompressionFormat compression in Enum.GetValues(typeof(CompressionFormat)))
{
    foreach (TransferMode transferMode in Enum.GetValues(typeof(TransferMode))
    {
        yield return new object[]
        {
                MessageVersion.Soap12WSAddressing10,
                System.ServiceModel.Channels.MessageVersion.CreateVersion(System.ServiceModel.EnvelopeVersion.Soap12, System.ServiceModel.Channels.AddressingVersion.WSAddressing10),
                compression,
                (System.ServiceModel.Channels.CompressionFormat)compression,
                transferMode,
                (System.ServiceModel.TransferMode)transferMode
        };
        yield return new object[]
        {
                MessageVersion.Soap12WSAddressingAugust2004,,
                System.ServiceModel.Channels.MessageVersion.Soap12WSAddressingAugust2004,
                compression,
                (System.ServiceModel.Channels.CompressionFormat)compression,
                transferMode,
                (System.ServiceModel.TransferMode)transferMode
        };
    }
}
```
I haven't tested this code so you might need to cast to int first when casting between the CoreWCF and the SM variations, but I think the above will work. Also GetValues returns an Array which has a signature of returning objects. I think the foreach does an implicit cast but it might need the iterator variable to be type object and then cast where it's used. I believe it should work as written, but there's a chance you might need to make these changes.

Comment (jonlouie) [reply src/CoreWCF.NetTcp/tests/CompressionFormatTests.cs:63]: Thanks for the suggestion- I've added it in my latest commit. Looks great, worked perfectly, and increased the test coverage from 8 to 24 cases.

**Comment by mconnew** [comment]
@jonlouie, this looks good now. Do you want to rebase and squash these down to a single commit before I merge or should I squash and merge it on github? The difference is it tends to loose the authorship if I squash and merge it on github. It ends up as part of the commit comment and not on the commit metadata so I'm not sure if it shows up on your github activity graph. I believe it doesn't.

**Comment by birojnayak** [comment]
> It looks like this automated test is failing for `Debug Windows_net5.0` and `Release Linux_net5.0`: CoreWCF.NetTcp.Tests.TransportWithMessageCredentialNetTCPTest.NetTCPRequestReplyWithTransportMessageEchoStringDemuxFailure
> 
> @mconnew Any idea if this is a flaky test? I'm not able to reproduce the failure locally.

it's flaky @jonlouie

**Comment by jonlouie** [comment]
Closing this PR to replace it with #627
